pep/003-dependency-namespaces-and-labels.md

```diff
@@ -0,0 +1,508 @@
+    * [Namespace Selector](#namespace-selector)
+    * [Name Selector](#name-selector)
+    * [Interface Selector](#interface-selector)
+  * [Use Case Solutions](#use-case-solutions)
+    * [Solution: Exclusive Resource](#solution-exclusive-resource)
+    * [Solution: Global Resource](#solution-global-resource)
+    * [Solution: Scoped Resource](#solution-scoped-resource)
+    * [Solution: User Provided Resource](#solution-user-provided-resource)
+    * [Solution: Polymorophic Resource](#solution-polymorphic-resource)
+  * [Porter Bundle Index](#porter-bundle-index)
+    * [List Indexed Bundles](#list-indexed-bundles)
+    * [Register a Bundle](#register-a-bundle)
+    * [Unregister a Bundle](#unregister-a-bundle)
+  * [Post-Install Dependency Management](#post-install-dependency-management)
+    * [Managed Dependency](#managed-dependency)
+      * [Detecting Unused Managed Dependencies](#detecting-unused-managed-dependencies)
+    * [Unmanaged Dependency](#unmanaged-dependency)
+  * [Dependency Resolution](#dependency-resolution)
+* [Implementation](#implementation)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Security Implications](#security-implications)
+* [Rejected Ideas](#rejected-ideas)
+* [Open Questions](#open-questions)
+
+## Abstract
+
+Support advanced dependency management use cases, such as shared dependencies, using existing installations, and satisfying a dependency with an interface, through the use of namespaces and labels.
+
+___
+
+## Motivation
+
+Initial Proposal: https://github.com/getporter/porter/discussions/1441
+
+When we initially implemented dependencies in Porter, we knew that there were a lot of different directions that we could take but we were missing strong use cases and experience with what a dependency looked and how it behaved with bundles.
+We implemented the most obvious scenario (wordpress requires its own mysql database) and decided to wait for further feedback before continuing.
+Now that we have been working with bundles for a while a set of use cases have evolved:
+
+### Use Case: Exclusive Resource
+
+> I need my own instance of my dependency
+
+When a bundle is run, new installations are created for its dependencies.
+For example, an application requires a dedicated redis cache per installation.
+This is the only supported use case by Porter today.
+
+
+### Use Case: Global Resource
+
+> I just need my dependency to be present
+
+I am writing the Porter Operator to work with Operator Lifecycle Manager (OLM).
+When I install the Porter Operator, OLM should be installed on the cluster.
+It should not be installed more than once.
+
+
+### Use Case: Scoped Resource
+
+> My micro-services share a common dependency
+
+My application has 4 micro-services and a shared postgres database.
+When I install the first service, a database should be created for it.
+When I install the other services, I want to share the database (and the connection string) between them.
+
```

Comment:
For instance, when using helm mixinm the release name could be used as identifier and helm will insure the idempotence of the operation

Reply:
### Use Case: Bundle to Bundle Reference
The current bundle is dependents on another bundle to allow cascading mechanism.
The parameters provided to the root bundle should be propagated to all inner bundles